examples/gno.land/p/demo/membstore/membstore.gno

```diff
@@ -0,0 +1,197 @@
+// with an initial member list
+func WithInitialMembers(members []Member) Option {
+	return func(store *MembStore) {
+		for _, m := range members {
+			memberAddr := m.Address.String()
+
+			// Check if the member already exists
+			if store.members.Has(memberAddr) {
+				panic(ufmt.Errorf("%s, %s", memberAddr, ErrAlreadyMember))
+			}
+
+			store.members.Set(memberAddr, m)
+			store.totalVotingPower += m.VotingPower
+		}
+	}
+}
+
+// WithDAOPkgPath initializes the member store
+// with a dao package path guard
+func WithDAOPkgPath(daoPkgPath string) Option {
+	return func(store *MembStore) {
+		store.daoPkgPath = daoPkgPath
+	}
+}
+
+// MembStore implements the dao.MembStore abstraction
+type MembStore struct {
+	daoPkgPath       string    // active dao pkg path, if any
+	members          *avl.Tree // std.Address -> Member
+	totalVotingPower uint64    // cached value for quick lookups
+}
+
+// NewMembStore creates a new member store
+func NewMembStore(opts ...Option) *MembStore {
+	m := &MembStore{
+		members:          avl.NewTree(), // empty set
+		daoPkgPath:       "",            // no dao guard
+		totalVotingPower: 0,
+	}
+
+	// Apply the options
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	return m
+}
+
+func (m *MembStore) AddMember(member Member) error {
+	if !m.isCallerDAORealm() {
+		return ErrNotGovDAO
+	}
+
+	// Check if the member exists
+	if m.IsMember(member.Address) {
+		return ErrAlreadyMember
+	}
+
+	// Add the member
+	m.members.Set(member.Address.String(), member)
+
+	// Update the total voting power
+	m.totalVotingPower += member.VotingPower
+
+	return nil
+}
+
+func (m *MembStore) UpdateMember(address std.Address, member Member) error {
+	if !m.isCallerDAORealm() {
+		return ErrNotGovDAO
+	}
+
+	// Get the member
+	oldMember, err := m.Member(address)
+	if err != nil {
+		return err
+	}
+
+	// Check if this is a removal request
+	if member.VotingPower == 0 {
+		m.members.Remove(address.String())
+
+		// Update the total voting power
+		m.totalVotingPower -= oldMember.VotingPower
+
+		return nil
+	}
+
+	// Check that the member wouldn't be
+	// overwriting an existing one
+	isAddressUpdate := address != member.Address
+	if isAddressUpdate && m.IsMember(member.Address) {
+		return ErrInvalidAddressUpdate
+	}
+
+	// Remove the old member info
+	// in case the address changed
+	if address != member.Address {
+		m.members.Remove(address.String())
+	}
+
+	// Save the new member info
+	m.members.Set(member.Address.String(), member)
+
+	// Update the total voting power
+	difference := member.VotingPower - oldMember.VotingPower
+	m.totalVotingPower += difference
+
+	return nil
+}
+
+func (m *MembStore) IsMember(address std.Address) bool {
+	_, exists := m.members.Get(address.String())
+
+	return exists
+}
+
+func (m *MembStore) Member(address std.Address) (Member, error) {
+	member, exists := m.members.Get(address.String())
+	if !exists {
+		return Member{}, ErrMissingMember
+	}
+
+	return member.(Member), nil
+}
+
```

Comment:
I had difficuties to understand `Members()` API. May be it would help it you keep the same idiom as Go slices: if m was a Go array, members(start, end uint64) would return m[start:end].

Reply:
This `offset, count` API (pagination) is a pattern we follow for most Gno contracts -- it comes from the AVL implementation. It's something we converged on some time ago.

It might be good to have an Effective Gno page on this @Kouteki

Reply:
> It might be good to have an Effective Gno page on this @Kouteki

https://github.com/gnolang/gno/issues/3044